Add README (#25)
openshift/operator-framework-tooling: README.md

# operator-framework-tooling

This repository contains tools for downstreaming OLM software

## OLMv0

The OLMv0 code is downstreamed into the [openshift/operator-framework-olm](https://github.com/openshift/operator-framework-olm) repository from the following upstream repositories:

* [operator-framework/api](https://github.com/operator-framework/api)
* [operator-framework/operator-lifecycle-manager](https://github.com/operator-framework/operator-lifecycle-manager) - the main repo
* [operator-framework/operator-registry](https://github.com/operator-framework/operator-registry)

> The `operator-lifecycle-manager` repository is the main repository, as it refers to the others. The its `go.mod` file determines what versions of `api` and `operator-registry` are used and downstreamed. Any integration between these three repositories _must_ be done upstream, as that will provide a clean downstreaming experience.

Reviewer: ```suggestion
The `operator-lifecycle-manager` repository is the main repository, as it refers to the others. Its `go.mod` file determines what versions of `api` and `operator-registry` are used and downstreamed. Any integration between these three repositories _must_ be done upstream, as that will provide a clean downstreaming experience.
```

The upstream repositories must also be `go mod tidy` clean.

The downstream repository is a "monorepo" consisting of the above reposities located in a `staging` directory. The upstream commits are cherry-picked with updating comments indicating the source repository and commit SHA.

## OLMv1

The OLMv1 code is downstreamed into the [openshift/operator-framework-operator-controller](https://github.com/openshift/operator-framework-operator-controller) repository from the following upstream repository:

* [operator-framework/operator-controller](https://github.com/operator-framework/operator-controller)

> There is only one repository that is downtreamed. The upstream commits are downstreamed as a direct mirror (i.e. commit SHAs remain the same) and then merged into the `main` branch.

Reviewer: I don't think this is true. There are downstream equivalents of catalogd + rukpak that are currently downstreamed by the syncer.

I believe the process is:
- Check for updates to catalogd + rukpak based on operator-controller go.mod
- If updates for catalogd + rukpak, sync them, don't sync operator-controller yet
- If no updates for catalogd + rukpak, sync operator-controller

Author: I haven't done any syncing for v1... but my understanding is that it's just `operator-controller` code that's merged, and catalogd + rukpak are vendored?

Author: I'm looking at the code, and it seems to merge all three, but only update go.mod of operator-controller if the others don't need changes. I'm not seeing how it's _not_ syncing operator-controller...

Reviewer: That part that I'm saying I don't think is true is that it _only_ syncs operator-controller

Reviewer: Sync PR of catalogd: https://github.com/openshift/operator-framework-catalogd/pull/46

Author: Right; I've updated the PR.

# operator-framework-tooling

This repository contains tools for downstreaming OLM software

## OLMv0

The OLMv0 code is downstreamed into the [openshift/operator-framework-olm](https://github.com/openshift/operator-framework-olm) repository from the following upstream repositories:

* [operator-framework/api](https://github.com/operator-framework/api)
* [operator-framework/operator-lifecycle-manager](https://github.com/operator-framework/operator-lifecycle-manager) - the main repo
* [operator-framework/operator-registry](https://github.com/operator-framework/operator-registry)

The `operator-lifecycle-manager` repository is the main repository, as it refers to the others. Its `go.mod` file determines what versions of `api` and `operator-registry` are used and downstreamed. Any integration between these three repositories _must_ be done upstream, as that will provide a clean downstreaming experience.

The upstream repositories must also be `go mod tidy` clean.

> The downstream repository is a "monorepo" consisting of the above reposities located in a `staging` directory. The upstream commits are cherry-picked with updating comments indicating the source repository and commit SHA.

Reviewer 2: ```suggestion
The downstream repository is a "monorepo" consisting all of the above reposities located in a `staging` directory. The upstream commits are cherry-picked with updating comments indicating the source repository and commit SHA.
```

## OLMv1

The OLMv1 code is downstreamed into the three separate repositories:

* [operator-framework/catalogd](https://github.com/operator-framework/catalogd) -> [openshift/operator-framework-catalogd](https://github.com/openshift/operator-framework-catalogd)
* [operator-framework/operator-controller](https://github.com/operator-framework/operator-controller) -> [openshift/operator-framework-operator-controller](https://github.com/openshift/operator-framework-operator-controller) - the main repo
* [operator-framework/rukpak](https://github.com/operator-framework/rukpak) -> [openshift/operator-framework-rukpak](https://github.com/openshift/operator-framework-rukpak)

The upstream commits are downstreamed as a direct mirror (i.e. commit SHAs remain the same) and then merged into the `main` branch.

However `operator-controller` has dependencies on the `catalogd` and `rukpak` repos, and requires that its `go.mod` is updated to reflect the downstream (openshift) repositories.
1. Use the upstream `go.mod` file in `operator-controller` to determine which commits of `catalogd` and `rukpak` are to be merged.

> 2. Synchronize all three downstream repos.

Reviewer 2: Is this a well-known procedure?  It it reflecting anything besides k8s bump coordination?

Author: Clarify the comment/question? I reviewed the code, and to the best of my knowledge, this is how it works.

Reviewer 2: "synchronize all three downstream repos" means...?

Author: I can add some stuff... but this was a non-trivial process!

* [operator-framework/rukpak](https://github.com/operator-framework/rukpak) -> [openshift/operator-framework-rukpak](https://github.com/openshift/operator-framework-rukpak)

The upstream commits are downstreamed as a direct mirror (i.e. commit SHAs remain the same) and then merged into the `main` branch.

However `operator-controller` has dependencies on the `catalogd` and `rukpak` repos, and requires that its `go.mod` is updated to reflect the downstream (openshift) repositories.
1. Use the upstream `go.mod` file in `operator-controller` to determine which commits of `catalogd` and `rukpak` are to be merged.
2. Synchronize all three upstream repo to their respective downstream repos.
3. If there are no changes to `catalogd` or `rukpak` (i.e. they've been merged), update the downstream `go.mod` file in `operator-controller` to reference those downstream repos.

Only if there are no outstanding merges for downstream `catalogd` or `rukpak`, is the `go.mod` file in downstream `operator-controller` updated.

Merging to downstream consists of:

> 1. Merging via merge commit upstream `main` branch, this overrides the existing `main` branch via `git merge --stategy=ours`. This keeps the upstream and downstream commits numbered witht he same SHA.

Reviewer 2: ```suggestion
1. Merging via merge commit upstream `main` branch, this overrides the existing `main` branch via `git merge --stategy=ours`. This keeps the upstream and downstream commits numbered with the same SHA.
```
